dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/TrackedEntitySecurityValidationHook.java

```diff
@@ -0,0 +1,122 @@
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import org.hisp.dhis.organisationunit.OrganisationUnit;
+import org.hisp.dhis.preheat.Preheat;
+import org.hisp.dhis.program.ProgramInstance;
+import org.hisp.dhis.security.Authorities;
+import org.hisp.dhis.trackedentity.TrackedEntityInstance;
+import org.hisp.dhis.trackedentity.TrackedEntityType;
+import org.hisp.dhis.tracker.preheat.PreheatHelper;
+import org.hisp.dhis.tracker.report.TrackerErrorCode;
+import org.hisp.dhis.tracker.TrackerIdentifier;
+import org.hisp.dhis.tracker.bundle.TrackerBundle;
+import org.hisp.dhis.tracker.domain.TrackedEntity;
+import org.hisp.dhis.tracker.report.TrackerErrorReport;
+import org.hisp.dhis.tracker.report.ValidationErrorReporter;
+import org.hisp.dhis.user.User;
+import org.springframework.stereotype.Component;
+
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.hisp.dhis.tracker.report.ValidationErrorReporter.newReport;
+
+/**
+ * @author Morten Svanæs <msvanaes@dhis2.org>
+ */
+@Component
+public class TrackedEntitySecurityValidationHook
+    extends AbstractTrackerValidationHook
+{
+
+    @Override
+    public int getOrder()
+    {
+        return 1;
+    }
+
+    @Override
+    public List<TrackerErrorReport> validate( TrackerBundle bundle )
+    {
+        ValidationErrorReporter reporter = new ValidationErrorReporter( bundle, this.getClass() );
+
+        User actingUser = bundle.getPreheat().getUser();
+
+        for ( TrackedEntity te : bundle.getTrackedEntities() )
+        {
+            reporter.increment();
+
+            TrackedEntityType entityType = getTrackedEntityType( bundle, te );
+
+            if ( entityType != null && !aclService.canDataWrite( actingUser, entityType ) )
+            {
+                reporter.addError( newReport( TrackerErrorCode.E1001 )
+                    .withObject( te )
+                    .addArg( actingUser ).addArg( entityType ) );
+            }
+
+            OrganisationUnit orgUnit = getOrganisationUnit( bundle, te );
+
+            if ( orgUnit != null && !organisationUnitService.isInUserSearchHierarchyCached( actingUser, orgUnit ) )
```

Comment:
I'm not sure, but I imagine it shouldnt be possible to create/update/delete data outside your capturescope; Which searchScope will most likely be. I am confirming this with @Bekkalizer .

Reply:
create/update/delete operations shouldn't be allowed outside the capture scope. search scope is for reading TEIs.

Reply:
@stian-sandvold @abyot Can you please clarify a little more on this. This code is copied verbatim from the old code. At least unless I have screwed up something...

Reply:
There are two main orgUnit "scopes", searchScope (All records you can see/search for) and captureScope (All records you can edit). I confirmed with Markus as well, that we dont want to allow users to edit data outside their captureScope. Imagine the captureScope as a small circle within the searchScope :)

It should be as easy as just changing the method-call form searchScope to captureScope.

Reply:
@stian-sandvold Please verify I am using the right method now.

Reply:
I don't see the changes here, but as long as you check that its captureScope and not SearchScope, it should be fine :)

Reply:
https://github.com/dhis2/dhis2-core/blob/6172ac4488498e0adda1ff66065f0e2514048977/dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/TrackedEntitySecurityValidationHook.java#L86